flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -293,140 +246,120 @@ public void close() throws Exception {
+	 * as asynchronous results can be processed.
+	 *
+	 * @param streamElement to add to the operator's queue
+	 * @throws InterruptedException if the current thread has been interrupted while yielding to mailbox
+	 * @return a handle that allows to set the result of the async computation for the given element.
+	 */
+	private ResultFuture<OUT> addToWorkQueue(StreamElement streamElement) throws InterruptedException {
+		assert(Thread.holdsLock(checkpointingLock));
 
-		try {
-			super.dispose();
-		} catch (InterruptedException interrupted) {
-			exception = interrupted;
+		pendingStreamElement = streamElement;
 
-			Thread.currentThread().interrupt();
-		} catch (Exception e) {
-			exception = e;
+		Optional<ResultFuture<OUT>> queueEntry;
+		while (!(queueEntry = queue.tryPut(streamElement)).isPresent()) {
+			mailboxExecutor.yield();
 		}
 
-		try {
-			stopResources(false);
-		} catch (InterruptedException interrupted) {
-			exception = ExceptionUtils.firstOrSuppressed(interrupted, exception);
+		pendingStreamElement = null;
 
-			Thread.currentThread().interrupt();
-		} catch (Exception e) {
-			exception = ExceptionUtils.firstOrSuppressed(e, exception);
-		}
+		return queueEntry.get();
+	}
 
-		if (exception != null) {
-			throw exception;
+	private void waitInFlightInputsFinished() throws InterruptedException {
+		assert(Thread.holdsLock(checkpointingLock));
+
+		while (!queue.isEmpty()) {
+			mailboxExecutor.yield();
 		}
 	}
 
 	/**
-	 * Close the operator's resources. They include the emitter thread and the executor to run
-	 * the queue's complete operation.
+	 * Batch output of all completed elements. Watermarks are always completed if it's their turn to be processed.
 	 *
-	 * @param waitForShutdown is true if the method should wait for the resources to be freed;
-	 *                           otherwise false.
-	 * @throws InterruptedException if current thread has been interrupted
+	 * <p>This method will be called from {@link #processWatermark(Watermark)} and from a mail processing the result
+	 * of an async function call.</p>
 	 */
-	private void stopResources(boolean waitForShutdown) throws InterruptedException {
-		emitter.stop();
-		emitterThread.interrupt();
-
-		executor.shutdown();
-
-		if (waitForShutdown) {
-			try {
-				if (!executor.awaitTermination(365L, TimeUnit.DAYS)) {
-					executor.shutdownNow();
-				}
-			} catch (InterruptedException e) {
-				executor.shutdownNow();
-
-				Thread.currentThread().interrupt();
-			}
-
-			/*
-			 * FLINK-5638: If we have the checkpoint lock we might have to free it for a while so
-			 * that the emitter thread can complete/react to the interrupt signal.
-			 */
-			if (Thread.holdsLock(checkpointingLock)) {
-				while (emitterThread.isAlive()) {
-					checkpointingLock.wait(100L);
+	private void outputCompletedElements() {
+		Collection<StreamElement> completed = queue.popCompleted();
```

Comment:
I'd like to avoid pulling the elements one-by-one. But your comment gave me the idea to simply pass the collector into the queue (which passes it to the entries). This will avoid creating and copying any intermediate results.

Reply:
What I've proposed is in fact would be very similar to the old code. The loop (traversal) overhead should be the same as it's now (with small "optimisation" that you would not check `.isWatermark()` for every output element but for every input element).
With this approach, I think it's possible to preserve more of old behavior and introduce back the `TimestampedCollector` here.

Reply:
Not sure what the benefit of the old code was, but I have improved the code even further. There is no need for the check anymore. So please check it one more time.

`TimestampedCollector` seems to be too limited at the moment (should actually delegate all `Output` methods). I could extend it in that way and use it (basically it's missing `emitWatermark` for our use case).